internal/jobcontainers/path.go

```diff
@@ -0,0 +1,228 @@
+// - Construct a list of searchable paths to find the application. This includes the standard Windows system paths
+// which are generally located at C:\Windows, C:\Windows\System32 and C:\Windows\System. If a working directory or path is specified
+// via the `workingDirectory` or `pathEnv` parameters then these will be appended to the paths to search from as well. The
+// searching logic is handled by the Windows API function `SearchPathW` which accepts a semicolon separated list of paths to search
+// in.
+// https://docs.microsoft.com/en-us/windows/win32/api/processenv/nf-processenv-searchpathw
+//
+// - If the commandline is quoted, simply grab whatever is in the quotes and search for this directly.
+// We don't try any other logic here, if the application can't be found from the quoted contents we return an error.
+//
+// - If the commandline is not quoted, we iterate over each possible application name by splitting the arguments and iterating
+// over them one by one while appending the last search each time until we either find a match or don't and return
+// an error. If we don't find the application on the first try, this means that the application name has a space in it
+// and we must adjust the commandline to add quotes around the application name.
+//
+// - If the application is found, we return the fullpath to the executable and the adjusted commandline (if needed).
+//
+// Examples:
+// - Input: "C:\Program Files\sub dir\program name"
+//  Search order:
+//  - C:\Program.exe
+//  - C:\Program Files\sub.exe
+//  - C:\Program Files\sub dir\program.exe
+//  - C:\Program Files\sub dir\program name.exe
+//  Returned commandline: "\"C:\Program Files\sub dir\program name\""
+//
+// - Input: "\"program name\""
+//  Search order:
+//  - program name.exe
+//  Returned commandline: "\"program name\"
+//
+// - Input: "\"program name\" -flags -for -program"
+//  Search order:
+//  - program.exe
+//  - program name.exe
+//  Returned commandline: "\"program name\" -flags -for -program"
+//
+// - Input: "\"C:\path\to\program name\""
+//  Search Order:
+//  - "C:\path\to\program name.exe"
+//  Returned commandline: "\"C:\path\to\program name""
+//
+// - Input: "C:\path\to\program"
+//  Search Order:
+//  - "C:\path\to\program.exe"
+//  Returned commandline: "C:\path\to\program"
+//
+// CreateProcess documentation: https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-createprocessa
+func getApplicationName(commandLine, workingDirectory, pathEnv string) (string, string, error) {
+	var (
+		searchPath string
+		result     string
+	)
+	// First we get the system paths concatenated with semicolons (C:\windows;C:\windows\system32;C:\windows\system;)
+	// and use this as the basis for the directories to search for the application.
+	systemPaths, err := getSystemPaths()
+	if err != nil {
+		return "", "", err
+	}
+
+	// If there's a working directory we should also add this to the list of directories to search.
```

Comment:
This case we might actually want to take out. Is working directory used for standard Windows containers?

Reply:
It's added to the searchable paths yes. This is how we search in the container images base directory for these also as we set the working dir to wherever the volume is mounted to.

Reply:
Gross. Okay.